### ISSUE TITLE
rust-analyzer spins forever "Fetching" while opening fuchsia workspace

### ISSUE THREAD

**Issue by ComputerDruid** [rust-lang/rust-analyzer#14276]
**rust-analyzer version**:

rust-analyzer version: 0.3.1426-standalone

**rustc version**: (eg. output of `rustc -V`)

rustc 1.69.0-nightly (9433ba639 2023-02-08)

**relevant settings**: (eg. client settings, or environment variables like `CARGO`, `RUSTC`, `RUSTUP_HOME` or `CARGO_HOME`)

Fuchsia uses `rust-project.json` (generated by the build system)

Unfortunately repro instructions require a lot of disk space because a fuchsia checkout is quite large. Nevertheless, you can get a fuchsia checkout by following the instructions at https://fuchsia.dev/fuchsia-src/get-started/get_fuchsia_source and running `./scripts fx set core.x64`, which will configure a small build and generate a rust-project.json.

I attached a debugger to the rust-analyzer and it seems to be spinning in `rust_analyzer::reload::ProjectFolders::new`, backtrace looks like:

```
Thread 2 (Thread 0x7f2c119ef6c0 (LWP 3198526) "LspServer"):
#0  0x0000558ad094d11a in core::hash::BuildHasher::hash_one ()
#1  0x0000558ad08ecf0b in hashbrown::rustc_entry::<impl hashbrown::map::HashMap<K,V,S,A>>::rustc_entry ()
#2  0x0000558ad0a4e0e7 in rust_analyzer::reload::ProjectFolders::new ()
#3  0x0000558ad0a7885a in rust_analyzer::reload::<impl rust_analyzer::global_state::GlobalState>::switch_workspaces ()
#4  0x0000558ad0a727d6 in rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::handle_task ()
#5  0x0000558ad0a6e479 in rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::handle_event ()
#6  0x0000558ad0a6d0e3 in rust_analyzer::main_loop::<impl rust_analyzer::global_state::GlobalState>::run ()
#7  0x0000558ad08f0de0 in rust_analyzer::main_loop::main_loop ()
#8  0x0000558ad0af6592 in rust_analyzer::run_server ()
#9  0x0000558ad0afbc26 in std::sys_common::backtrace::__rust_begin_short_backtrace ()
#10 0x0000558ad0abf1cf in core::ops::function::FnOnce::call_once{{vtable.shim}} ()
#11 0x0000558ad0b5e155 in alloc::boxed::{impl#45}::call_once<(), dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global> () at library/alloc/src/boxed.rs:2000
#12 alloc::boxed::{impl#45}::call_once<(), alloc::boxed::Box<dyn core::ops::function::FnOnce<(), Output=()>, alloc::alloc::Global>, alloc::alloc::Global> () at library/alloc/src/boxed.rs:2000
#13 std::sys::unix::thread::{impl#2}::new::thread_start () at library/std/src/sys/unix/thread.rs:108
#14 0x00007f2c11b33fd4 in start_thread (arg=<optimized out>) at ./nptl/pthread_create.c:442
#15 0x00007f2c11bb466c in clone3 () at ../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
```

I suspect this was probably caused by #14218

**Comment by ComputerDruid** [comment]
If it helps, I'm attaching the `rust-project.json` from my build. 

[rust-project.json](https://github.com/rust-lang/rust-analyzer/files/10915150/rust-project.json.txt)

**Comment by ComputerDruid** [comment]
Because there's no output when it's hung like this, I'm not 100% sure how to minimize this to not require a fuchsia checkout, but I'm happy to volunteer some time to repro with some logging if that is helpful

**Comment by Veykril** [comment]
Oh, looks like I the include clearing is missing in the loop so it never exits when there is overlap.

**Comment by Veykril** [comment]
No actually we do clear them on iteration, I don't see the immediate problem there.

**Comment by ComputerDruid** [comment]
So far, I've attached some debugging to the top of the loop in question, it spins through iterations of the loop rapidly; by iteration 100, the overlap map seems to be stable at:

<details>
  <summary><code>[crates/rust-analyzer/src/reload.rs:544] &overlap_map = {</code></summary>

```
    939: [
        1074,
        1695,
    ],
    338: [
        340,
        340,
    ],
    716: [
        717,
    ],
    273: [
        274,
    ],
    583: [
        1611,
    ],
    360: [
        1325,
    ],
    264: [
        265,
    ],
    887: [
        1672,
    ],
    218: [
        351,
        351,
    ],
    342: [
        345,
        345,
    ],
    373: [
        1670,
    ],
    308: [
        355,
        355,
    ],
    305: [
        880,
    ],
    271: [
        272,
    ],
    355: [
        360,
        364,
        374,
        374,
        374,
        384,
        384,
    ],
    1300: [
        1702,
    ],
    1074: [
        1075,
        1075,
        1076,
        1076,
        1077,
        1077,
        1078,
        1078,
        1079,
        1079,
        1080,
        1080,
        1081,
        1081,
        1082,
        1082,
        1083,
        1083,
        1084,
        1084,
        1085,
        1085,
        1086,
        1086,
        1087,
        1087,
        1088,
        1088,
        1089,
        1089,
        1090,
        1090,
        1091,
        1091,
        1092,
        1092,
        1093,
        1093,
        1094,
        1094,
        1095,
        1095,
        1096,
        1096,
        1097,
        1097,
        1098,
        1098,
        1099,
        1099,
        1100,
        1100,
        1101,
        1101,
        1102,
        1102,
        1103,
        1103,
        1104,
        1104,
        1105,
        1105,
        1106,
        1106,
        1107,
        1107,
        1108,
        1108,
        1109,
        1109,
        1110,
        1110,
        1111,
        1111,
        1112,
        1112,
        1113,
        1113,
        1114,
        1114,
        1115,
        1115,
        1116,
        1116,
        1117,
        1117,
        1118,
        1118,
        1119,
        1119,
        1120,
        1120,
        1121,
        1121,
        1122,
        1122,
        1123,
        1123,
        1124,
        1124,
        1125,
        1125,
        1126,
        1126,
        1127,
        1127,
        1128,
        1128,
        1129,
        1129,
        1130,
        1130,
        1131,
        1131,
        1132,
        1132,
        1133,
        1133,
        1134,
        1134,
        1135,
        1135,
        1136,
        1136,
        1137,
        1137,
        1138,
        1138,
        1139,
        1139,
        1140,
        1140,
        1141,
        1141,
        1142,
        1142,
        1143,
        1143,
        1144,
        1144,
        1145,
        1145,
        1146,
        1146,
        1147,
        1147,
        1148,
        1148,
        1149,
        1149,
        1150,
        1150,
        1151,
        1151,
        1152,
        1152,
        1153,
        1153,
        1154,
        1154,
        1155,
        1155,
        1156,
        1156,
        1157,
        1157,
        1158,
        1158,
        1159,
        1159,
        1160,
        1160,
        1161,
        1161,
        1162,
        1162,
        1163,
        1163,
        1164,
        1164,
        1165,
        1165,
        1166,
        1166,
        1167,
        1167,
        1168,
        1168,
        1168,
        1169,
        1169,
        1170,
        1170,
        1170,
        1171,
        1171,
        1172,
        1172,
        1173,
        1173,
        1174,
        1174,
        1175,
        1175,
        1176,
        1176,
        1177,
        1177,
        1178,
        1178,
        1179,
        1179,
        1180,
        1180,
        1181,
        1181,
        1182,
        1182,
        1183,
        1183,
        1184,
        1184,
        1184,
        1185,
        1185,
        1186,
        1186,
        1186,
        1187,
        1187,
        1188,
        1188,
        1189,
        1189,
        1190,
        1190,
        1191,
        1191,
        1192,
        1192,
        1193,
        1193,
        1194,
        1194,
        1195,
        1195,
        1196,
        1196,
        1197,
        1197,
        1198,
        1198,
        1199,
        1199,
        1200,
        1200,
        1201,
        1201,
        1202,
        1202,
        1203,
        1203,
        1204,
        1204,
        1205,
        1205,
        1206,
        1206,
        1207,
        1207,
        1208,
        1208,
        1209,
        1209,
        1210,
        1210,
        1211,
        1211,
        1212,
        1212,
        1213,
        1213,
        1214,
        1214,
        1215,
        1215,
        1216,
        1216,
        1217,
        1217,
        1218,
        1218,
        1219,
        1219,
        1220,
        1220,
        1221,
        1221,
        1222,
        1222,
        1223,
        1223,
        1224,
        1224,
        1225,
        1225,
        1226,
        1226,
        1227,
        1227,
        1228,
        1228,
        1229,
        1229,
        1230,
        1230,
        1231,
        1231,
        1232,
        1232,
        1233,
        1233,
        1234,
        1234,
        1235,
        1235,
        1236,
        1236,
        1237,
        1237,
        1238,
        1238,
        1239,
        1239,
        1240,
        1240,
        1241,
        1241,
        1242,
        1242,
        1243,
        1243,
        1244,
        1244,
        1245,
        1245,
        1246,
        1246,
        1247,
        1247,
        1248,
        1248,
        1249,
        1249,
        1250,
        1250,
        1251,
        1251,
        1252,
        1252,
        1253,
        1253,
        1254,
        1254,
        1255,
        1255,
        1256,
        1256,
        1257,
        1257,
        1258,
        1258,
        1259,
        1259,
        1260,
        1260,
        1261,
        1261,
        1262,
        1262,
        1263,
        1263,
        1264,
        1264,
        1265,
        1265,
        1266,
        1266,
        1267,
        1267,
        1268,
        1268,
        1269,
        1269,
        1270,
        1270,
        1271,
        1271,
        1272,
        1272,
        1273,
        1273,
        1274,
        1274,
        1275,
        1275,
        1276,
        1276,
        1277,
        1277,
        1278,
        1278,
        1279,
        1279,
        1280,
        1280,
        1281,
        1281,
        1282,
        1282,
        1283,
        1283,
        1284,
        1284,
        1285,
        1285,
        1286,
        1286,
        1287,
        1287,
        1288,
        1288,
        1289,
        1289,
        1290,
        1290,
        1291,
        1291,
        1292,
        1292,
        1293,
        1293,
        1294,
        1294,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1695,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
        1738,
    ],
    727: [
        728,
        728,
        728,
        728,
    ],
    315: [
        355,
    ],
    219: [
        308,
        315,
        355,
        355,
        360,
        364,
        374,
        384,
    ],
    1301: [
        1703,
    ],
    260: [
        261,
    ],
    669: [
        1651,
    ],
    1298: [
        1701,
    ],
    257: [
        258,
    ],
    632: [
        633,
    ],
}
```
</details>

When I went to narrow down why stuff wasn't getting removed from the path, I finally spotted the bug, on later iterations of the loop, it edits the wrong index in the roots, the correct logic is:

```patch
diff --git a/crates/rust-analyzer/src/reload.rs b/crates/rust-analyzer/src/reload.rs
index fca242d91..28d37f568 100644
--- a/crates/rust-analyzer/src/reload.rs
+++ b/crates/rust-analyzer/src/reload.rs
@@ -527,7 +527,7 @@ impl ProjectFolders {
             // maps include paths to indices of the corresponding root
             let mut include_to_idx = FxHashMap::default();
             // Find and note down the indices of overlapping roots
-            for (idx, root) in roots.iter().filter(|it| !it.include.is_empty()).enumerate() {
+            for (idx, root) in roots.iter().enumerate().filter(|(_, it)| !it.include.is_empty()) {
                 for include in &root.include {
                     match include_to_idx.entry(include) {
                         Entry::Occupied(e) => {
```

In other words, the indices we were getting back were offset because we were enumerating the filtered list.

I'll upload a PR momentarily
